node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/DynamicProvisioningMaintainer.java

```diff
@@ -0,0 +1,161 @@
+            NodeList nodes = nodeRepository().list();
+
+            updateProvisioningNodes(nodes, lock);
+            convergeToCapacity(nodes);
+        }
+    }
+
+    void updateProvisioningNodes(NodeList nodes, Mutex lock) {
+        Map<String, Node> provisionedHostsByHostname = nodes.state(Node.State.provisioned).nodeType(NodeType.host)
+                .asList().stream()
+                .collect(Collectors.toMap(Node::hostname, Function.identity()));
+
+        Map<Node, Set<Node>> nodesByProvisionedParent = nodes.asList().stream()
+                .filter(node -> node.parentHostname().map(provisionedHostsByHostname::containsKey).orElse(false))
+                .collect(Collectors.groupingBy(
+                        node -> provisionedHostsByHostname.get(node.parentHostname().get()),
+                        Collectors.toSet()));
+
+        nodesByProvisionedParent.forEach((host, children) -> {
+            try {
+                List<Node> updatedNodes = hostProvisioner.provision(host, children);
+                nodeRepository().write(updatedNodes, lock);
+            } catch (IllegalArgumentException | IllegalStateException e) {
+                log.log(Level.INFO, "Failed to provision " + host.hostname() + ": " + Exceptions.toMessageString(e));
+            } catch (FatalProvisioningException e) {
+                log.log(Level.SEVERE, "Failed to provision " + host.hostname() + ", failing out the host recursively", e);
+                // Fail out as operator to force a quick redeployment
+                nodeRepository().failRecursively(
+                        host.hostname(), Agent.operator, "Failed by HostProvisioner due to provisioning failure");
+            } catch (RuntimeException e) {
+                log.log(Level.WARNING, "Failed to provision " + host.hostname() + ", will retry in " + interval(), e);
+            }
+        });
+    }
+
+    void convergeToCapacity(NodeList nodes) {
+        List<Node> emptyHosts = getAllocatableHosts(nodes);
+        List<NodeResources> preProvisionCapacity = preprovisionCapacityFlag.value().stream()
+                .flatMap(cap -> {
+                    NodeResources resources = new NodeResources(cap.getVcpu(), cap.getMemoryGb(), cap.getDiskGb(), 1);
+                    return IntStream.range(0, cap.getCount()).mapToObj(i -> resources);
+                })
+                .sorted(NodeResourceComparator.memoryDiskCpuOrder())
+                .collect(Collectors.toList());
+
+        int hostI = 0;
+        int capacityI = 0;
+        while (hostI < emptyHosts.size() && capacityI < preProvisionCapacity.size()) {
+            if (NodePrioritizer.ALLOCATABLE_HOST_STATES.contains(emptyHosts.get(hostI).state()) &&
+                    emptyHosts.get(hostI).flavor().resources().satisfies(preProvisionCapacity.get(capacityI))) {
+                emptyHosts.remove(hostI);
+                preProvisionCapacity.remove(capacityI);
```

Comment:
Note: List typically have a rather poor remove(0) performance

Reply:
Yes, but these will typically be tiny lists, probably less than 5 elements.